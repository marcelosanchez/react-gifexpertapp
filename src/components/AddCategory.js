import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ( { setCategories } ) => {

    const [inputValue, setInputValue] = useState('');  // Si no se pone nada, va como undefined

    const handleInputChange = ( e ) => {
        // console.log(e.target.value)
        setInputValue(e.target.value)
    }

    const handleSubmit = (e) => {
        e.preventDefault();

        if( inputValue.trim().length > 1 ) {  // Controla que no se envie vacio
            setCategories( cats => [ inputValue, ...cats ] );
            setInputValue('');  // Limpia el input
        }

    };

    return (
        <>
            <form onSubmit={handleSubmit}>
                <input 
                    type="text" 
                    value={ inputValue }
                    onChange={ handleInputChange }
                />
            </form>
        </>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}