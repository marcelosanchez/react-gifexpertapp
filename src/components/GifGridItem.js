import React from 'react'

export const GifGridItem = ( { id, title, url } ) => {

    return (
        <div className='card animate__fadeIn animate__delay-2s'>
            <img src={ url } alt={ title } />
            <p>{ title }</p>
        </div>
    )
}
