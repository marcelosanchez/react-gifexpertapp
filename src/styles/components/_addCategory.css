._addCategory__form {
    background-color: #fff;
    border-top: 1px solid #f5f6f8;
    display: flex;
    height: 100%;
    padding: 0;
    margin: 0;
    padding-top: 20px;
    width: 100%;
}

._addCategory__input {
    background: transparent url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='16' height='16' class='bi bi-search' viewBox='0 0 16 16'%3E%3Cpath d='M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z'%3E%3C/path%3E%3C/svg%3E") no-repeat 13px center;
    background-color: #f1f3f6;
    border: none;
    border-radius: 5px;
    margin-bottom: 10px;
    outline: none;
    padding: 10px;
    padding-left: 40px;
}

._addCategory__gifGridContainer {
    padding: 0px;
}