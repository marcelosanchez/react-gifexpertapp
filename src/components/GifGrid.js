import React, { useState, useEffect } from 'react'
import { useFetchGfs } from '../hooks/useFetchGifs'
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ( { category } ) => {

    // const [images, setImages] = useState([])
    const { data:images, loading } = useFetchGfs( category );
    // console.log(images, loading);

    // getGifs();

    return (
        <>
            <h3 className='animate__animated animate__fadeIn'>{ category }</h3>
            { loading && <p className='animate__animated animate__flash'>'Loading...'</p> }
            <div className='card-grid'>
                { 
                    images.map( img => 
                        <GifGridItem 
                            key={img.id} 
                            { ...img } 
                        />
                    ) 
                }
            </div>
        </>
    )
}
