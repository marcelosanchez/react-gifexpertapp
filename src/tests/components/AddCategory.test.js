import React from 'react';
import '@testing-library/jest-dom';

import { shallow } from 'enzyme';
import { AddCategory } from '../../components/AddCategory';

describe('Pruebas del componente <AddCategory/>', () => { 
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories = { setCategories }/>);

    beforeEach( () => {
        // Todo lo que necesito que se resetee, en cada prueba
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories = { setCategories }/>);
    } );

    test('Debe mostrarse correctamente', () => { 
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe cambiar la caja de texto', () => {
        const input = wrapper.find('input');
        const value = `Hola mundo`;

        input.simulate('change', { target: { value } });  // Las llaves son para simular el evento
        
        expect( wrapper.find('p').text().trim() ).toBe( value );
    });

    test('No debe postear la informacion onsubmit', () => {
        wrapper.find('form').simulate('submit', { preventDefault: () => {} })
        expect( setCategories ).not.toHaveBeenCalled()  // no se llama la funcion
    });

    test('Debe llamar el setCategories y limpiar la caja de texto', () => {
        const value = `Hola mundo`;
        const input = wrapper.find('input');
        
        // 1. simular el imputChange
        input.simulate('change', { target: { value } });  // Las llaves son para simular el evento
        // 2. simular el submit
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });
        // 3. setCategories se debe haber llamado
        expect( setCategories ).toHaveBeenCalled();
        expect( setCategories ).toHaveBeenCalledTimes( 1 );  // Numero de veces que se llama la funcion
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) );  // Que el parametro enviado sea una funcion

        // 4. el valor del input de estar como un string vacio
        console.log(input.text)
        expect( input.prop('value') ).toBe('');
    });
})