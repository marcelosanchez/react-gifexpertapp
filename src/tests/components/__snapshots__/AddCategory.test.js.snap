// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente <AddCategory/> Debe mostrarse correctamente 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <p />
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </form>
</Fragment>
`;
